blockstore/apps/bundles/data/store.py

```diff
@@ -0,0 +1,222 @@
+    """
+    bundle_uuid = attr.ib()
+
+    # Map of paths to FileInfo objects
+    files = attr.ib(type=dict)
+
+    # All our various links and their extended dependencies
+    dependencies = attr.ib(type=DependencyMapping)
+
+class BundleJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, BundleVersionKey):
+            return str(obj)
+        elif isinstance(obj, FileInfo):
+            return [
+                obj.path,
+                obj.public,
+                obj.size,
+                obj.hash.hexdigest(),
+            ]
+        elif isinstance(obj, pathlib.Path):
+            return str(obj)
+        elif isinstance(obj, BundleVersionSummary):
+            return {
+                'bundle_uuid': str(obj.bundle_uuid),
+                'files': obj.files,
+                'dependencies': obj.dependencies,
+            }
+        return json.JSONEncoder.default(self, obj)
+
+
+class BundleDataStore():
+    """
+    This is the interface for storing the actual files that make up a Bundle.
+
+    The responsibilities of this class are:
+    * Create a new Snapshot from a set of files.
+    * Generate URLs for resources within a Bundle.
+    * Validate Links.
+
+    # Important note about Versions/Snapshots:
+
+    Numbered versions don't exist at this level -- versioning here is entirely
+    based on content, so basically some kind of hash. BundleVersion is what
+    manages numbered versioning and points to an entry in BundleDataStore. To
+    make that distinction clearer, we'll say that BundleDataStore creates
+    Snapshots, which BundleVersion will point to.
+
+    This also helps us to mitigate the concurrency issue. Creating a snapshot
+    might take a long time (say to upload all the assets). If we were assigning
+    a version_num at the start of that process, we might run into a race
+    condition with a process that starts a few seconds later, because they'll
+    both see the database reflecting the same state. But if BundleStore creates
+    only hash-versioned Snapshots, then it can do all that work and the only
+    time version_num is read and incremented for a new entry is when
+    BundleDataStore emits a signal that BundleVersion catches.
+
+    # Extensibility
+
+    Basic: We use Django's File Storage API, which means that one extension
+    point is the Storages layer. This is the layer we'd plug into to start,
+    using django-storages for things like S3 and equivalents, and some in-memory
+    backing store for running tests.
+
+    Advanced: It's possible that one day we'll want to swap out the entire
+    BundleDataStore to adopt a radically different versioning store like git
+    (+LFS). In that case, we'd create a new class that implements the same
+    public BundleStore interface. To keep the option open, we'd want to make
+    sure that things going in and out of BundleStore don't leak storage
+    implementation details.
+    """
```

Comment:
I've been mucking around with version modeling all day, and I think I'm landing on a different separation than what was originally put in the proposal. It basically involves pushing version number understanding up to the models.py level, while keeping the stuff below that in the BundleDataStore purely content-hash based (the BundleVersion model bridges the two).

Haven't quite figured out the implications on dependency calculation. It might have some positive implications for content removal (i.e. the separation allows us to remove content snapshots without necessarily removing the BundleVersion rows).

Anyhow, a lot to munch on, and it's probably pushed this work out a day. But I think the separation will be good for both separation of concerns and operational concurrency issues.

Reply:
Have you considered including the content address of the previous snapshot in the summary to create a chain and maintaining ordering at the snapshot level? If we did that then the version "numbers" would become derived data (though we may still want to keep them in the model for efficient reads).

Reply:
It's crossed my mind, but I'm wary of adding more version control features. Right now, a new snapshot can be created without knowing anything about the previous snapshot, and a split where two entries have the same parent manifests in two versions being created. It's not smart, but it's simple.

I'm definitely open to it down the line, especially if it's needed to make drafts work well. But I'd rather hold off until we have a concrete use for tracking that level.